plugin-server/src/main/ingestion-queues/session-recording/services/session-manager-v3.ts

```diff
@@ -0,0 +1,472 @@
+        const manager = new SessionManagerV3(serverConfig, s3Client, context)
+        await mkdir(context.dir, { recursive: true })
+
+        try {
+            const bufferMetadata: SessionManagerBufferContext = JSON.parse(
+                await readFile(path.join(context.dir, 'metadata.json'), 'utf-8')
+            )
+            manager.buffer = {
+                context: bufferMetadata,
+                fileStream: manager.createFileStreamFor(path.join(context.dir, BUFFER_FILE_NAME)),
+            }
+        } catch (error) {
+            // Indicates no buffer metadata file or it's corrupted
+            status.error('🧨', '[session-manager] failed to read buffer metadata', {
+                ...context,
+                error,
+            })
+        }
+
+        status.info('📦', '[session-manager] started new manager', {
+            ...manager.context,
+            ...(manager.buffer?.context ?? {}),
+        })
+        return manager
+    }
+
+    private async save(): Promise<void> {
+        if (this.buffer) {
+            await writeFile(this.file('metadata.json'), JSON.stringify(this.buffer?.context), 'utf-8')
+        } else {
+            await unlink(this.file('metadata.json'))
+        }
+    }
+
+    private async getFlushFiles(): Promise<string[]> {
+        return (await readdir(this.context.dir)).filter((file) => file.endsWith(FLUSH_FILE_EXTENSION))
+    }
+
+    private captureException(error: Error, extra: Record<string, any> = {}): void {
+        captureException(error, {
+            extra: { ...this.context, ...extra },
+            tags: { teamId: this.context.teamId, sessionId: this.context.sessionId },
+        })
+    }
+
+    private captureMessage(message: string, extra: Record<string, any> = {}): void {
+        const context = this.context
+        captureMessage(message, {
+            extra: { ...context, ...extra },
+            tags: { teamId: context.teamId, sessionId: context.sessionId },
+        })
+    }
+
+    public async add(message: IncomingRecordingMessage): Promise<void> {
+        if (this.destroying) {
+            return
+        }
+
+        try {
+            const buffer = this.getOrCreateBuffer()
+            const messageData = convertToPersistedMessage(message)
+            const start = message.events.at(0)?.timestamp
+            const end = message.events.at(-1)?.timestamp ?? start
+
+            if (!start || !end) {
+                captureMessage("[session-manager]: can't set events range from message without events summary", {
+                    extra: { message },
+                })
+                return
+            }
+
+            buffer.context.eventsRange = {
+                firstTimestamp: minDefined(start, buffer.context.eventsRange?.firstTimestamp) ?? start,
+                lastTimestamp: maxDefined(end, buffer.context.eventsRange?.lastTimestamp) ?? end,
+            }
+
+            const content = JSON.stringify(messageData) + '\n'
```

Comment:
random thought... what happens if the `messageData` has `\n` in it already. e.g. poorly encoded console log data or something

Reply:
Interesting. Could be the source of some existing bugs (that behaviour is the same in the other consumer.